### PR TITLE
factorio: give advice on how to retain source tarball

### PR DESCRIPTION
**Pull request by bmillwood**
[edit: following discussion I changed approach of this PR, but retaining the original PR description for posterity)

This option artificially makes the source tarball a runtime dependency of the final derivation by symlinking it into the share directory. This is to ensure it doesn't get GC'd, which otherwise can mean you need to manually redownload it every time some unrelated dependency of the package updates.

I've been using this personally for a while. I think it's a pretty janky solution, so I'd be very interested to hear more idiomatic ways to achieve the same thing, but I'm also happy for this to be merged if other folks are ok with it.

Note that even though I only really care about this option in the `needsAuth` cases, it seemed least confusing to just make it always do the same thing.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
    - I only tried this with the expansion client, but I don't expect the others to be any different
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
You probably want to set keep-outputs/keep-derivations for nix instead.

**Comment by SuperSandro2000** [comment]
> I can't target that at a specific package, right?

nope

> I think the case for using it for factorio in particular is unusually strong, since if you can't get the credentials working it has to be download manually.

Why can't you get the credentials to work? That's more the main problem here.

Also just running their standalone download with steam-run also works pretty well.

**Comment by Artturin** [comment]
You could just do `system.extraDependencies [ factorio.src ];`

**Comment by bmillwood** [comment]
> > I think the case for using it for factorio in particular is unusually strong, since if you can't get the credentials working it has to be download manually.
> 
> Why can't you get the credentials to work? That's more the main problem here.

To be honest after my first attempt I didn't really look into it, since the package immediately recommended another way to me, so I just did that instead. But thinking about it now, I still don't want to redownload the tar even if I do get it working automatically, because:
1. it takes time to download, it's nearly 4G,
2. I use copy-on-write snapshotting backups, so every time i delete and redownload it, I lose sharing between my backups, so I store another copy of it, wasting another 4G.

I think if we recommend the prefetch method in the error text, we should also try to make it work nicely.

> Also just running their standalone download with steam-run also works pretty well.

I mean sure, but with that attitude why package it at all?

> You could just do `system.extraDependencies [ factorio.src ];`

ah, thanks. I thought I had tried something like that when I first came up with this trick, but I knew less about nix back then and I guess didn't try the right things.

I've pushed a version that just gives this advice, although I think it would be improved if I also had advice for people using nix without nixos.

File: pkgs/by-name/fa/factorio/package.nix [hunk 1]
```diff
@@ -70,6 +70,15 @@ let
       nix-prefetch-url file://\''$HOME/Downloads/factorio_\''${releaseType}_x64_\''${version}.tar.xz --name factorio_\''${releaseType}_x64-\''${version}.tar.xz
 
     Note the ultimate "_" is replaced with "-" in the --name arg!
+
+    If you go this route you might want to tell nix to explicitly hold on to the
+    source tarball. Otherwise it could get GC'd from the nix store and you'd
```
Comment (bjornfor) [inline pkgs/by-name/fa/factorio/package.nix:75]: nix -> Nix

File: pkgs/by-name/fa/factorio/package.nix [hunk 2]
```diff
@@ -70,6 +70,15 @@ let
...
+    have to redownload it next time the package wants to rebuild to use a newer
+    dependency. E.g. if you're using nixos:
```
Comment (bjornfor) [inline pkgs/by-name/fa/factorio/package.nix:77]: nixos -> NixOS

File: pkgs/by-name/fa/factorio/package.nix (hunk 1 above)
Comment (bmillwood) [reply pkgs/by-name/fa/factorio/package.nix:75]: did you miss the nix in the previous line, or are you drawing a deliberate distinction between the program `nix` in lowercase and the project `Nix` in uppercase or something?

Comment (bmillwood) [reply pkgs/by-name/fa/factorio/package.nix:75]: (btw if you use GitHub's "add a suggestion" button, it's slightly more convenient to credit you with these changes)

**Comment by bmillwood** [comment]
@SuperSandro2000 what do you think about this purely-suggestion version?

File: pkgs/by-name/fa/factorio/package.nix (hunk 1 above)
Comment (bjornfor) [reply pkgs/by-name/fa/factorio/package.nix]: Yes, I missed the first "nix". Yes, I think we should differentiate between the `nix` CLI and the Nix package manager. No, I don't think the first "nix" refers to the CLI.

(I don't need/want credit for such small changes -- I'd rather have the changes squashed -- but thanks for the reminder that it's a more precise way to suggest changes.)

**Review by SuperSandro2000 — APPROVED** [review]
LGTM

**Pull request merged**